buffer.go

```diff
@@ -11,142 +10,103 @@ import (
+	// PutRecordsLimit is the maximum number of records allowed for a PutRecords request.
+	putRecords int
 
-// PutRecordsSizeLimit is the maximum allowed size per PutRecords request.
-var PutRecordsSizeLimit int = 5 * 1024 * 1024 // 5MB
+	// PutRecordsSizeLimit is the maximum allowed size per PutRecords request.
+	putRecordsSize int
 
-type recordBuffer struct {
-	client   *kinesis.Kinesis
-	pKeyTmpl *template.Template
-	input    *kinesis.PutRecordsInput
-	count    int
-	byteSize int
-	mutex    sync.Mutex
+	// RecordSizeLimit is the maximum allowed size per record.
+	recordSize int
 }
 
-func newRecordBuffer(client *kinesis.Kinesis, streamName string) (*recordBuffer, error) {
-	pKeyTmpl, err := compileTmpl("KINESIS_PARTITION_KEY_TEMPLATE")
-	if err != nil {
-		return nil, err
-	}
-
-	input := &kinesis.PutRecordsInput{
-		StreamName: aws.String(streamName),
-		Records:    make([]*kinesis.PutRecordsRequestEntry, 0),
-	}
-
-	return &recordBuffer{
-		client:   client,
-		pKeyTmpl: pKeyTmpl,
-		input:    input,
-	}, nil
-}
+const (
+	PutRecordsLimit     int = 500
+	PutRecordsSizeLimit int = 5 * 1024 * 1024 // 5MB
+	RecordSizeLimit     int = 1 * 1024 * 1024 // 1MB
+)
 
-type recordSizeLimitError struct {
-	caller string
-	length int
+type buffer struct {
+	count    int
+	byteSize int
+	pKeyTmpl *template.Template
+	input    kinesis.PutRecordsInput
+	limits   *limits
 }
 
-func (e *recordSizeLimitError) Error() string {
-	return fmt.Sprintf("%s: log data byte size (%d) is over the limit.", e.caller, e.length)
+func newBuffer(tmpl *template.Template, sn string) *buffer {
+	return &buffer{
+		pKeyTmpl: tmpl,
+		input: kinesis.PutRecordsInput{
+			StreamName: aws.String(sn),
+			Records:    make([]*kinesis.PutRecordsRequestEntry, 0),
+		},
+		limits: &limits{
+			putRecords:     PutRecordsLimit,
+			putRecordsSize: PutRecordsSizeLimit,
+			recordSize:     RecordSizeLimit,
+		},
+	}
 }
 
-// Add fills the buffer with new data, or flushes it if one of the limit
-// has hit.
-func (r *recordBuffer) Add(m *router.Message) error {
-	data := m.Data
-	dataLen := len(data)
+func (b *buffer) add(m *router.Message) error {
+	dataLen := len(m.Data)
 
 	// This record is too large, we can't submit it to kinesis.
-	if dataLen > RecordSizeLimit {
-		return &recordSizeLimitError{
-			caller: "recordBuffer.Add",
-			length: dataLen,
-		}
-	}
-
-	defer r.mutex.Unlock()
-	r.mutex.Lock()
-
-	// Adding this event would make our request have too many records. Flush first.
-	if r.count+1 > PutRecordsLimit {
-		err := r.Flush()
-		if err != nil {
-			return err
-		}
-	}
-
-	// Adding this event would make our request too large. Flush first.
-	if r.byteSize+dataLen > PutRecordsSizeLimit {
-		err := r.Flush()
-		if err != nil {
-			return err
-		}
+	if dataLen > b.limits.recordSize {
+		return ErrRecordTooBig
 	}
 
-	// Partition key
-	pKey, err := executeTmpl(r.pKeyTmpl, m)
+	pKey, err := executeTmpl(b.pKeyTmpl, m)
 	if err != nil {
-		return err
+		panic(err)
```

Comment:
I talk about this in another comment, but I think we really don't want to crash, because that will result in lost log lines for other adapters that were behaving correctly. We should definitely be loud about the error, but crashing and restarting won't do much good since we'll just keep crashing most likely.